gunicorn/sock.py

```diff
@@ -203,10 +204,27 @@ def create_sockets(conf, log, fds=None):
 
     return listeners
 
-
 def close_sockets(listeners, unlink=True):
     for sock in listeners:
         sock_name = sock.getsockname()
         sock.close()
         if unlink and _sock_type(sock_name) is UnixSocket:
             os.unlink(sock_name)
+
+def ssl_context(conf):
+    context = conf.ssl_context(conf)
+    if context is None:
+        context = ssl.SSLContext(conf.ssl_version)
```

Comment:
@tsaarni I wanted to check whether you considered using [ssl.create_default_context(...)](https://docs.python.org/3/library/ssl.html?highlight=sslkeylogfile#ssl.create_default_context) here instead of constructing directly.

From docs:
- https://github.com/python/cpython/blob/5765eaa13654e5f812a286700da7d6b8e144da0e/Lib/ssl.py#L740-L746
> The protocol and settings may change anytime without prior
          deprecation. The values represent a fair balance between maximum
          compatibility and security.
          
- https://docs.python.org/3/library/ssl.html?highlight=sslkeylogfile#ssl.create_default_context
> The settings are chosen by the [ssl](https://docs.python.org/3/library/ssl.html?highlight=sslkeylogfile#module-ssl) module, and usually represent a higher security level than when calling the [SSLContext](https://docs.python.org/3/library/ssl.html?highlight=sslkeylogfile#ssl.SSLContext) constructor directly.

Looks like Python is encouraging use here if you want to adopt over-time the baseline secure default-options and features for ssl, with a leaning towards most-secure-defaults.  When changes are made in future versions of Python to disable or enable features, they might be made here.

So I think a trade-off between no-unplanned-changes seen by Gunicorn, vs auto-adopt best defaults, which are then overridable.

Interesting also that when the key-logging option was added in https://github.com/python/cpython/commit/c7f7069e77c58e83b847c0bfe4d5aadf6add2e68 , it was added to `create_default_context()` and not the ctor.

WDYT?

Reply:
Hey @javabrett, thanks for your comments! 

> I wanted to check whether you considered using ssl.create_default_context(...) here instead of constructing directly.
> ...
> Looks like Python is encouraging use here if you want to adopt over-time the baseline secure default-options and features for ssl, with a leaning towards most-secure-defaults. When changes are made in future versions of Python to disable or enable features, they might be made here.

Using that would mean removing the existing `--ssl_version` config option since `ssl.create_default_context()` does not accept that option, unlike `SSLContext` does.  People who care about TLS details are likely aware that `--ssl_version` has not fully worked anymore with recent versions of Python, but I was bit unsure about touching that. I assumed the PR might be easier to accept if it is fully backwards compatible, especially since there has not been that much development activity in the project recently. But I do understand if some kind of deprecation process might be wanted to reduce old baggage!

Looking at the current implementation of `ssl.create_default_context()` there is not much there at this point besides of the  `SSLKEYLOGFILE` support.  But of course in _future_ this may change and in that case it wuold be nice to pick up secure defaults from there.

> I did some ergonomics testing on the new config setting and it highlighted another idea I had earlier
> ...
> I don't have to know/copy/maintain the standard Gunicorn-specific standard configuration logic of the default SSLContext, as I (optionally) obtain the base object from conf._create_default_ssl_context() (name could probably use some work). Currently this knowledge has to be copied from example_config.py or sock.py, which might later change.

Good idea!  I added a commit in this PR that adds the functionality but I'd like to propose following:  It looked to me like `Config` class has not been used for method like this, so I just added a factory function `default_ssl_context_factory` as the second parameter to the hook function. I've also updated the example code accordingly. I hope this has the same usability improvement as adding capability to `Config` to construct `SSLContexts`.

Reply:
i kepe that discussio open. I would like also go to defaults. A release is planned this week, i will try to add that change myself. If it's a quick change for you feel free to dit by now :)

Reply:
@benoitc How should the deprecation of the existing config option be done? Should it raise some error that points out the new way? Are we ok with non-backwards compatible change that might potentially break applications?

Reply:
good question. That will be a major version, so breaking changes will need at least to be documented. If we want to make it cleaner maybe we can raise a Deprecation warning when the option is used and just ignore it?